wp/lib/bap_wp/src/z3_utils.ml

```diff
@@ -100,6 +100,113 @@ let get_z3_name (map : Constr.z3_expr Var.Map.t) (name : string) (fmt : Var.t ->
         else
           None)
 
+(** Sexp.t helper constructors for smtlib constructs *)
+let define_fun name args retsort body =
+  let args = List.map ~f:(fun (v,sort) -> Sexp.List [v; sort]) args in
+  Sexp.List [Sexp.Atom "define-fun"; Sexp.Atom name; Sexp.List args; retsort; body]
+let synonym name ret body = define_fun name [] ret body
+let define_sort (name : string) s = Sexp.List [
+ Sexp.Atom "define-sort"; Sexp.Atom name; Sexp.List []; s]
+let sexp_of_sort (s : Z3.Sort.sort) : Sexp.t = Sexp.of_string (Z3.Sort.to_string s)
+
+(* [make_arg_synonyms] constructs smtlib formula for the arguments of a subroutine.
+   This make synonyms that for high level c-like names ot low level locations *)
+let make_arg_synonyms
+  ?postfix:(postfix = "")
+  (sub : Sub.t)
+  (init_var_map : Expr.expr Var.Map.t)
+  (var_map : Expr.expr Var.Map.t) : Sexp.t list =
+  let args = Term.enum arg_t sub |> Seq.to_list in
+  List.concat_map args ~f:(fun arg ->
+    let cname = Arg.lhs arg in
+    let syn prefix z3_reg =
+      let sort = Sexp.of_string (Z3.Sort.to_string (Z3.Expr.get_sort z3_reg)) in
+      synonym (prefix ^ Var.name cname ^ postfix) sort
+      (Sexp.Atom  (Expr.to_string z3_reg))
+    in
+    let reg = Arg.rhs arg in
+    match reg with
+    | Var reg -> begin
+         match Arg.intent arg with
+           | Some In   -> [syn "init_" (Var.Map.find_exn init_var_map reg)]
+           | Some Out  -> [syn "" (Var.Map.find_exn var_map reg)]
+           | Some Both -> [syn "init_" (Var.Map.find_exn init_var_map reg);
+                           syn "" (Var.Map.find_exn var_map reg)]
+           | None -> []
+    end
+    | _ -> [] (* We don't make synonym unless c variable is just held in a bap variable *)
+    )
+
+(** [args_prelude_compare] builds smtlib synonyms in a comparison for the arguments of
+    two compared subroutines. *)
+let args_prelude_compare
+  (main_sub1 : Sub.t)
```

Comment:
This might need to change on rebase: there's a new type `'a code` which represents a pair of an `'a term` and an `env`.

Reply:
(it's in `compare.mli`)